<a href="https://colab.research.google.com/github/yaramirezte/TelecomX_LATAM/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [178]:
import pandas as pd

In [120]:
url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

In [121]:
datos_TelecomX = pd.read_json(url)

In [122]:
datos_TelecomX.sample(10)

customerID Churn                                           customer  \
6682  9190-MFJLN   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
698   0993-OSGPT    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
1587  2256-YLYLP    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
1154  1636-NTNCO    No  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
2412  3372-CDXFJ    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
456   0644-OQMDK    No  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
105   0178-SZBHO    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
5860  8015-IHCGW    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1219  1740-CSDJP   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2192  3069-SSVSN    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                                  phone  \
6682     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
698      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1587     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1154  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
2412    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
456      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
105     {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
5860    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
1219  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
2192  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
6682  {'InternetService': 'Fiber optic', 'OnlineSecu...   
698   {'InternetService': 'DSL', 'OnlineSecurity': '...   
1587  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1154  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2412  {'InternetService': 'No', 'OnlineSecurity': 'N...   
456   {'InternetService': 'Fiber optic', 'OnlineSecu...   
105   {'InternetService': 'Fiber optic', 'OnlineSecu...   
5860  {'InternetService': 'Fiber optic', 'OnlineSecu...   
1219  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2192  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
6682  {'Contract': 'Month-to-month', 'PaperlessBilli...  
698   {'Contract': 'One year', 'PaperlessBilling': '...  
1587  {'Contract': 'Two year', 'PaperlessBilling': '...  
1154  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2412  {'Contract': 'One year', 'PaperlessBilling': '...  
456   {'Contract': 'Month-to-month', 'PaperlessBilli...  
105   {'Contract': 'Month-to-month', 'PaperlessBilli...  
5860  {'Contract': 'Two year', 'PaperlessBilling': '...  
1219  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2192  {'Contract': 'One year', 'PaperlessBilling': '...

#🔧 Transformación

In [123]:
datos_TelecomX['customer'].iloc[0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

El dataset presenta una estructura JSON anidada, donde varias columnas contienen diccionarios con información detallada del cliente, servicios contratados y datos de facturación.

In [124]:
datos_TelecomX.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [125]:
datos_TelecomX.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [30]:
datos_TelecomX.describe()

customerID Churn                                           customer  \
count         7267  7267                                               7267   
unique        7267     3                                                891   
top     9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
freq             1  5174                                                223   

                                                 phone  \
count                                             7267   
unique                                               3   
top     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
freq                                              3495   

                                                 internet  \
count                                                7267   
unique                                                129   
top     {'InternetService': 'No', 'OnlineSecurity': 'N...   
freq                                                 1581   

                                                  account  
count                                                7267  
unique                                               6931  
top     {'Contract': 'Month-to-month', 'PaperlessBilli...  
freq                                                    6

In [126]:
datos_TelecomX.isnull().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [127]:
datos_TelecomX.isna().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [128]:
datos_TelecomX.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [129]:
datos_TelecomX.map(type).head()

customerID          Churn        customer           phone  \
0  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
1  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
2  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
3  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
4  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   

         internet         account  
0  <class 'dict'>  <class 'dict'>  
1  <class 'dict'>  <class 'dict'>  
2  <class 'dict'>  <class 'dict'>  
3  <class 'dict'>  <class 'dict'>  
4  <class 'dict'>  <class 'dict'>

In [130]:
# pd.json_normalize() Recorre cada diccionario, entra dentro de los diccionarios anidados, saca cada clave interna, crea columnas nuevas usando notación con puntos
# datos_TelecomX.to_dict(orient="records") - Convierte el DataFrame en una lista de diccionarios, uno por fila.

datos_TelecomX_normalizados = pd.json_normalize(datos_TelecomX.to_dict(orient="records"))


In [131]:
datos_TelecomX_normalizados.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
2386  3329-WDIOK   Yes          Female                       0   
7178  9867-XOBQA    No          Female                       0   
5618  7669-LCRSD    No            Male                       0   
925   1304-SEGFY    No          Female                       0   
3293  4588-YBNIB   Yes          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
2386               No                  No                3                 No   
7178               No                 Yes               21                 No   
5618              Yes                 Yes               12                Yes   
925               Yes                  No               72                Yes   
3293               No                  No                1                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
2386    No phone service                      DSL  ...                   Yes   
7178    No phone service                      DSL  ...                   Yes   
5618                  No                      DSL  ...                    No   
925                  Yes              Fiber optic  ...                    No   
3293                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
2386                        No                   No                  Yes   
7178                       Yes                  Yes                  Yes   
5618                        No                   No                  Yes   
925                        Yes                   No                  Yes   
3293       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
2386                      Yes   Month-to-month                      Yes   
7178                       No         Two year                       No   
5618                       No   Month-to-month                      Yes   
925                       Yes         Two year                      Yes   
3293      No internet service   Month-to-month                      Yes   

     account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
2386      Electronic check                   50.60                 155.35  
7178          Mailed check                   53.15                 1183.2  
5618          Mailed check                   56.30                 628.65  
925       Electronic check                   97.85                 6841.3  
3293          Mailed check                   19.60                   19.6  

[5 rows x 21 columns]

In [132]:
datos_TelecomX_normalizados.duplicated().sum()

np.int64(0)

In [135]:
datos_TelecomX_normalizados['account.Contract'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [134]:
datos_TelecomX_normalizados['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [46]:
datos_TelecomX_normalizados['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [136]:
datos_TelecomX_normalizados.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total'],
      dtype='object')

In [48]:
datos_TelecomX_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [49]:
datos_TelecomX_normalizados.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

El conjunto de datos presenta información demográfica de los clientes características de los servicios contratados y datos contractuales y financieros. Tambien se consultó el diccionario de datos proporcionado. A partir de esta revisión, se identificó la variable Churn como objetivo, y se seleccionaron como variables relevantes aquellas relacionadas con el perfil del cliente, los servicios contratados y las condiciones del contrato.

In [137]:
df = datos_TelecomX_normalizados.copy()

In [138]:
for col in df.select_dtypes(include="object").columns:
  df[col] = df[col].str.strip()

In [139]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [140]:
(df['Churn'] == '').sum()

np.int64(224)

In [141]:
churn_vacios = df[df['Churn'] == '']

In [142]:
churn_vacios

customerID Churn customer.gender  customer.SeniorCitizen  \
30    0047-ZHDTW                Female                       0   
75    0120-YZLQA                  Male                       0   
96    0154-QYHJU                  Male                       0   
98    0162-RZGMZ                Female                       1   
175   0274-VVQOQ                  Male                       1   
...          ...   ...             ...                     ...   
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
30                 No                  No               11                Yes   
75                 No                  No               71                Yes   
96                 No                  No               29                Yes   
98                 No                  No                5                Yes   
175               Yes                  No               65                Yes   
...               ...                 ...              ...                ...   
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
30                   Yes              Fiber optic  ...                    No   
75                    No                       No  ...   No internet service   
96                    No                      DSL  ...                   Yes   
98                    No                      DSL  ...                   Yes   
175                  Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
30                          No                   No                   No   
75         No internet service  No internet service  No internet service   
96                          No                  Yes                   No   
98                          No                  Yes                   No   
175                        Yes                   No                  Yes   
...                        ...                  ...                  ...   
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
30                         No   Month-to-month                      Yes   
75        No internet service         Two year                      Yes   
96                         No         One year                      Yes   
98                         No   Month-to-month                    

In [143]:
churn_vacios[['customer.tenure', 'account.Charges.Total', 'account.Charges.Monthly']].head()

customer.tenure account.Charges.Total  account.Charges.Monthly
30                11                 929.3                    79.00
75                71                1355.1                    19.90
96                29                1696.2                    58.75
98                 5                287.85                    59.90
175               65               6792.45                   103.15

In [144]:
# eliminar los registros con df['Churn'] vacios
df = df[df['Churn'] != '']

In [56]:
# df['Churn'] = df['Churn'].str.capitalize()

In [145]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

Se identificaron 224 registros con la variable objetivo Churn vacía. Al explorar estos clientes se observó que poseen historial de facturación y antigüedad variable (customer.tenure), pero no es posible inferir su estado de evasión. Por este motivo, estos registros fueron eliminados del conjunto de datos, asegurando la integridad de la variable objetivo.

In [146]:
df['account.Charges.Total'].dtype

dtype('O')

In [147]:
df['account.Charges.Total']= pd.to_numeric(df['account.Charges.Total'], errors='coerce')

In [148]:
df['account.Charges.Total'].dtype

dtype('float64')

In [149]:
df.isnull().sum()

customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64

In [150]:
df.isna().sum()

customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64

In [151]:
df.loc[df['account.Charges.Total'].isnull(), ['account.Charges.Total', 'customer.tenure', 'account.Charges.Monthly']]

account.Charges.Total  customer.tenure  account.Charges.Monthly
975                     NaN                0                    56.05
1775                    NaN                0                    20.00
1955                    NaN                0                    61.90
2075                    NaN                0                    19.70
2232                    NaN                0                    20.25
2308                    NaN                0                    25.35
2930                    NaN                0                    73.35
3134                    NaN                0                    25.75
3203                    NaN                0                    52.55
4169                    NaN                0                    80.85
5599                    NaN                0                    19.85

In [152]:
df.loc[df['customer.tenure'] == 0, 'account.Charges.Total'] = 0

In [153]:
df['account.Charges.Total'].isnull().sum()

np.int64(0)

In [154]:
df.isna().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

Se identificaron 11 valores nulos en la variable account.Charges.Total. Al analizar estos registros se comprobó que corresponden a clientes con antigüedad (customer.tenure) igual a cero, quienes aún no han generado cargos acumulados. Por este motivo, los valores nulos fueron actualizados con el valor 0, garantizando la coherencia de la información financiera.

In [155]:
df.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
4879  6680-NENYN   Yes          Female                       0   
5376  7340-KEFQE    No          Female                       0   
6165  8441-SHIPE    No          Female                       0   
1011  1422-DGUBX    No            Male                       0   
1339  1932-UEDCX    No            Male                       1   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
4879               No                  No               43                Yes   
5376              Yes                 Yes               51                 No   
6165               No                  No               67                Yes   
1011              Yes                  No               32                Yes   
1339              Yes                  No               33                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
4879                 Yes              Fiber optic  ...                   Yes   
5376    No phone service                      DSL  ...                    No   
6165                 Yes                       No  ...   No internet service   
1011                 Yes              Fiber optic  ...                    No   
1339                 Yes              Fiber optic  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
4879                       Yes                   No                  Yes   
5376                        No                  Yes                   No   
6165       No internet service  No internet service  No internet service   
1011                        No                  Yes                   No   
1339                        No                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
4879                      Yes         Two year                      Yes   
5376                       No         Two year                       No   
6165      No internet service         Two year                      Yes   
1011                       No         One year                      Yes   
1339                      Yes   Month-to-month                      Yes   

          account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
4879           Electronic check                  104.60                4759.85  
5376  Bank transfer (automatic)                   34.20                1782.00  
6165  Bank transfer (automatic)                   25.60                1790.35  
1011           Electronic check                   79.25                2619.15  
1339  Bank transfer (automatic)                  100.60                3270.25  

[5 rows x 21 columns]

In [156]:
df['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [157]:
df['internet.OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [158]:
# Consultar
df.loc[df['internet.InternetService'] == 'No',['internet.InternetService', 'internet.OnlineSecurity', 'internet.OnlineBackup']]


internet.InternetService internet.OnlineSecurity internet.OnlineBackup
20                         No     No internet service   No internet service
23                         No     No internet service   No internet service
24                         No     No internet service   No internet service
27                         No     No internet service   No internet service
28                         No     No internet service   No internet service
...                       ...                     ...                   ...
7250                       No     No internet service   No internet service
7252                       No     No internet service   No internet service
7256                       No     No internet service   No internet service
7257                       No     No internet service   No internet service
7261                       No     No internet service   No internet service

[1526 rows x 3 columns]

Se verificaron las variables relacionadas con servicios de internet y se observó que, para los clientes sin servicio de internet (InternetService = "No"), las variables de servicios adicionales toman el valor "No internet service". Esta codificación es consistente y representa correctamente la ausencia de dichos servicios, por lo que no se realizaron modificaciones en estas columnas.

In [159]:
df.duplicated().sum()

np.int64(0)

In [160]:
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7043 non-null   object 
 1   Churn                      7043 non-null   object 
 2   customer.gender            7043 non-null   object 
 3   customer.SeniorCitizen     7043 non-null   int64  
 4   customer.Partner           7043 non-null   object 
 5   customer.Dependents        7043 non-null   object 
 6   customer.tenure            7043 non-null   int64  
 7   phone.PhoneService         7043 non-null   object 
 8   phone.MultipleLines        7043 non-null   object 
 9   internet.InternetService   7043 non-null   object 
 10  internet.OnlineSecurity    7043 non-null   object 
 11  internet.OnlineBackup      7043 non-null   object 
 12  internet.DeviceProtection  7043 non-null   object 
 13  internet.TechSupport       7043 non-null   object 
 1

customerID Churn customer.gender  customer.SeniorCitizen  \
count         7043  7043            7043             7043.000000   
unique        7043     2               2                     NaN   
top     9995-HOTOH    No            Male                     NaN   
freq             1  5174            3555                     NaN   
mean           NaN   NaN             NaN                0.162147   
std            NaN   NaN             NaN                0.368612   
min            NaN   NaN             NaN                0.000000   
25%            NaN   NaN             NaN                0.000000   
50%            NaN   NaN             NaN                0.000000   
75%            NaN   NaN             NaN                0.000000   
max            NaN   NaN             NaN                1.000000   

       customer.Partner customer.Dependents  customer.tenure  \
count              7043                7043      7043.000000   
unique                2                   2              NaN   
top                  No                  No              NaN   
freq               3641                4933              NaN   
mean                NaN                 NaN        32.371149   
std                 NaN                 NaN        24.559481   
min                 NaN                 NaN         0.000000   
25%                 NaN                 NaN         9.000000   
50%                 NaN                 NaN        29.000000   
75%                 NaN                 NaN        55.000000   
max                 NaN                 NaN        72.000000   

       phone.PhoneService phone.MultipleLines internet.InternetService  ...  \
count                7043                7043                     7043  ...   
unique                  2                   3                        3  ...   
top                   Yes                  No              Fiber optic  ...   
freq                 6361                3390                     3096  ...   
mean                  NaN                 NaN                      NaN  ...   
std                   NaN                 NaN                      NaN  ...   
min                   NaN                 NaN                      NaN  ...   
25%                   NaN                 NaN                      NaN  ...   
50%                   NaN                 NaN                      NaN  ...   
75%                   NaN                 NaN                      NaN  ...   
max                   NaN                 NaN                      NaN  ...   

       internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
count                   7043                      7043                 7043   
unique                     3                         3                    3   
top                       No                        No                   No   
freq                    3088                      3095                 3473   
mean                     NaN                       NaN                  NaN   
std                      NaN                       NaN                  NaN   
min                      NaN                       NaN                  NaN   
25%                      NaN                       NaN                  NaN   
50%                      NaN                       NaN                  NaN   
75%                      NaN                       NaN                  NaN   
max                      NaN                       NaN                  NaN   

       internet.StreamingTV internet.StreamingMovies account.Contract  \
count                  7043                     7043             7043   
unique                    3                        3                3   
top                      No                       No   Month-to-month   
freq                   2810                     2785             3875   
mean                    NaN                      NaN              NaN   
std                     NaN                      NaN              NaN   
min          

In [164]:
# Crear la columna Cuentas_Diarias
df.loc[:,'account.Charges.Daily'] = (df['account.Charges.Monthly'] / 30).round(2)

In [165]:
df[['account.Charges.Daily', 'account.Charges.Monthly']].head(5)

account.Charges.Daily  account.Charges.Monthly
0                   2.19                     65.6
1                   2.00                     59.9
2                   2.46                     73.9
3                   3.27                     98.0
4                   2.80                     83.9

Se creó la columna Cuentas_Diarias dividiendo la facturación mensual entre 30 días, proporcionando una métrica de gasto diario por cliente. Esta nueva variable permite analizar de manera más granular el comportamiento de los clientes y detectar patrones de evasión en función del gasto diario.

In [166]:
df.head(5)

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.DeviceProtection  \
0                      DSL  ...                        No   
1                      DSL  ...                        No   
2              Fiber optic  ...                       Yes   
3              Fiber optic  ...                       Yes   
4              Fiber optic  ...                        No   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   
3                   No                  Yes                      Yes   
4                  Yes                  Yes                       No   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         One year                      Yes          Mailed check   
1   Month-to-month                       No          Mailed check   
2   Month-to-month                      Yes      Electronic check   
3   Month-to-month                      Yes      Electronic check   
4   Month-to-month                      Yes          Mailed check   

  account.Charges.Monthly account.Charges.Total  account.Charges.Daily  
0                    65.6                593.30                   2.19  
1                    59.9                542.40                   2.00  
2                    73.9                280.85                   2.46  
3                    98.0               1237.85                   3.27  
4                    83.9                267.40                   2.80  

[5 rows x 22 columns]

In [167]:
for col in df.select_dtypes(include='object').columns:
    print(col, "→", df[col].unique(), "\n")

customerID → ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH'] 

Churn → ['No' 'Yes'] 

customer.gender → ['Female' 'Male'] 

customer.Partner → ['Yes' 'No'] 

customer.Dependents → ['Yes' 'No'] 

phone.PhoneService → ['Yes' 'No'] 

phone.MultipleLines → ['No' 'Yes' 'No phone service'] 

internet.InternetService → ['DSL' 'Fiber optic' 'No'] 

internet.OnlineSecurity → ['No' 'Yes' 'No internet service'] 

internet.OnlineBackup → ['Yes' 'No' 'No internet service'] 

internet.DeviceProtection → ['No' 'Yes' 'No internet service'] 

internet.TechSupport → ['Yes' 'No' 'No internet service'] 

internet.StreamingTV → ['Yes' 'No' 'No internet service'] 

internet.StreamingMovies → ['No' 'Yes' 'No internet service'] 

account.Contract → ['One year' 'Month-to-month' 'Two year'] 

account.PaperlessBilling → ['Yes' 'No'] 

account.PaymentMethod → ['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)'] 



In [168]:
# Convertir valores textuales en binarios
binarias = ['customer.Partner', 'customer.Dependents', 'phone.PhoneService',
            'phone.MultipleLines', 'internet.OnlineSecurity', 'internet.OnlineBackup',
            'internet.DeviceProtection', 'internet.TechSupport', 'internet.StreamingTV',
            'internet.StreamingMovies', 'account.PaperlessBilling']

In [169]:
# se transforma
for col in binarias:
    df.loc[:,col] = df[col].map({'Yes': 1, 'No': 0,'No internet service': 0, 'No phone service': 0 })

In [170]:
df.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
3394  4706-AXVKM   Yes          Female                       1   
2196  3071-MVJCD    No            Male                       0   
1230  1760-CAZHT    No            Male                       0   
1583  2252-JHJGE    No            Male                       0   
299   0431-APWVY   Yes          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
3394                0                   0               11                  1   
2196                1                   1               72                  1   
1230                0                   1               25                  1   
1583                0                   0               61                  1   
299                 1                   1               28                  1   

     phone.MultipleLines internet.InternetService  ...  \
3394                   1              Fiber optic  ...   
2196                   1                      DSL  ...   
1230                   0                       No  ...   
1583                   1              Fiber optic  ...   
299                    0              Fiber optic  ...   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
3394                         0                    0                    0   
2196                         1                    1                    1   
1230                         0                    0                    0   
1583                         1                    0                    1   
299                          0                    0                    1   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
3394                        1   Month-to-month                        1   
2196                        0         Two year                        0   
1230                        0   Month-to-month                        0   
1583                        1         One year                        0   
299                         1         One year                        1   

          account.PaymentMethod account.Charges.Monthly account.Charges.Total  \
3394    Credit card (automatic)                    84.8                906.85   
2196  Bank transfer (automatic)                    82.3               5815.15   
1230               Mailed check                    20.8                460.20   
1583           Electronic check                   104.0               6363.45   
299   Bank transfer (automatic)                    90.1               2598.95   

      account.Charges.Daily  
3394                   2.83  
2196                   2.74  
1230                   0.69  
1583                   3.47  
299                    3.00  

[5 rows x 22 columns]

In [171]:
df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})

In [188]:
import numpy as np

# Suponiendo que tu columna de gasto diario se llama 'Cuentas_Diarias'
# Verificamos mínimos y máximos
min_val = df['account.Charges.Daily'].min()
max_val = df['account.Charges.Daily'].max()

In [194]:
# Creamos 5 bins equidistantes entre min y max
bins = np.linspace(min_val, max_val, 6)  # 6 límites = 5 segmentos
labels = ['Very Low', 'Low', 'Medium', 'High', 'Very High']

In [195]:
# Crear columna segmentada
df['Daily_Segment'] = pd.cut(
    df['account.Charges.Daily'],
    bins=bins,
    labels=labels,
    include_lowest=True  # Incluye el valor mínimo en el primer bin
)


In [191]:
# Verificamos el resultado
print(df[['account.Charges.Daily', 'Daily_Segment']].head(10))
print("\nDistribución de segmentos:")
print(df['Daily_Segment'].value_counts())

   account.Charges.Daily Daily_Segment
0                   2.19        Medium
1                   2.00        Medium
2                   2.46        Medium
3                   3.27          High
4                   2.80          High
5                   2.31        Medium
6                   3.66     Very High
7                   2.82          High
8                   1.61           Low
9                   3.02          High

Distribución de segmentos:
Daily_Segment
High         1827
Very Low     1801
Medium       1371
Very High    1033
Low          1011
Name: count, dtype: int64


In [198]:
# analizar churn por segmento
#df.groupby('Daily_Segment', observed=True)['Churn'].mean()

#📊 Carga y análisis

#📄Informe final